spec/smile-identity-core/address_verification_spec.rb

```diff
@@ -0,0 +1,166 @@
+        expect(headers['smileid-source-sdk-version']).to eq(SmileIdentityCore::VERSION)
+        expect(headers['smileid-request-signature']).to eq(signature[:signature])
+      end
+    end
+
+    describe 'unsupported country codes' do
+      before do
+        response = Typhoeus::Response.new(
+          code: 400,
+          body: '{"code":"2413","error":"\"country\" must be one of [NG, ZA]","success":false}',
+        )
+        Typhoeus.stub("#{connection.instance_variable_get(:@url)}/async-verify-address").and_return(response)
+      end
+
+      it 'raises a runtime error for unsupported countries' do
+        invalid_payload = payload.merge(country: 'DD')
+        expect { connection.submit_job(invalid_payload) }
+          .to raise_error(RuntimeError,
+            /400:\s*{\s*"code":"2413",\s*"error":"\\"country\\" must be one of \[NG, ZA\]",\s*"success":false\s*}/)
+      end
+    end
+
+    describe '#submit_job' do
+      it 'submits the job' do
+        allow(connection).to receive(:submit_job).and_return(address_verification_response)
+        expect(connection.submit_job(payload)).to eq(address_verification_response)
+      end
+    end
+
+    describe '#query_job_status' do
+      it 'queries the job status' do
+        allow(connection).to receive(:query_job_status).and_return(address_verification_response)
+        expect(connection.query_job_status).to eq(address_verification_response)
+      end
+    end
+
+    describe 'payload validation' do
+      it 'raises an error when country is missing' do
+        invalid_payload = payload.except(:country)
+        expect { connection.submit_job(invalid_payload) }.to raise_error(RuntimeError)
+      end
+
+      it 'raises an error when address is missing' do
+        invalid_payload = payload.except(:address)
+        expect { connection.submit_job(invalid_payload) }.to raise_error(RuntimeError)
+      end
+
+      it 'raises an error when callback_url is missing' do
+        invalid_payload = payload.except(:callback_url)
+        expect { connection.submit_job(invalid_payload) }.to raise_error(RuntimeError)
+      end
+
+      context 'when the country is NG' do
+        it 'raises an error when utility_number is missing' do
+          invalid_payload = payload.except(:utility_number)
+          expect { connection.submit_job(invalid_payload) }.to raise_error(RuntimeError)
+        end
+
+        it 'raises an error when utility_provider is missing' do
+          invalid_payload = payload.except(:utility_provider)
+          expect { connection.submit_job(invalid_payload) }.to raise_error(RuntimeError)
+        end
+      end
+
+      context 'when the country is ZA' do
+        it 'raises an error when id_number is missing' do
+          za_payload = payload.merge(country: 'ZA').except(:utility_number, :utility_provider, :id_number)
+          expect { connection.submit_job(za_payload) }.to raise_error(RuntimeError)
```

Comment:
I would check for the error message and see if its `id_number is required` and I would do the same for other required fields per country